File: bench/cross-language/fannkuch/fannkuch.zig
const std = @import("std");

const N = 10;

pub fn main() !void {
    const stdout_file = std.fs.File.stdout();
    var buf: [4096]u8 = undefined;
    var w = stdout_file.writer(&buf);

    var perm: [N]usize = undefined;
    var perm1: [N]usize = undefined;
    var count: [N]usize = undefined;
    var max_flips: i32 = 0;
    var checksum: i32 = 0;
    var r: usize = N;
    var perm_count: usize = 0;

    for (0..N) |i| {
        perm1[i] = i;
    }

    outer: while (true) {
        while (r > 1) {
            count[r - 1] = r;
            r -= 1;
        }

        @memcpy(&perm, &perm1);

        // Count flips
        var flips: i32 = 0;
        var k: usize = perm[0];
        while (k != 0) {
            // Reverse first k+1 elements
            std.mem.reverse(usize, perm[0 .. k + 1]);
            flips += 1;
            k = perm[0];
        }

        if (flips > max_flips) max_flips = flips;
        if (perm_count % 2 == 0) {
            checksum += flips;
        } else {
            checksum -= flips;
        }
        perm_count += 1;

        // Next permutation
        while (true) {
            if (r == N) break :outer;
            const p0 = perm1[0];
            for (0..r) |i| {
                perm1[i] = perm1[i + 1];
            }
            perm1[r] = p0;
            count[r] -= 1;
            if (count[r] > 0) break;
            r += 1;
        }
    }

    try w.interface.print("{d}\nPfannkuchen({d}) = {d}\n", .{ checksum, N, max_flips });
    try w.interface.flush();
}

File: bench/cross-language/fannkuch/fannkuch.go
package main

import "fmt"

const N = 10

func main() {
	perm := make([]int, N)
	perm1 := make([]int, N)
	count := make([]int, N)
	maxFlips := 0
	checksum := 0
	r := N
	permCount := 0

	for i := 0; i < N; i++ {
		perm1[i] = i
	}

	for {
		for r > 1 {
			count[r-1] = r
			r--
		}

		copy(perm, perm1)

		// Count flips
		flips := 0
		k := perm[0]
		for k != 0 {
			// Reverse first k+1 elements
			for i, j := 0, k; i < j; i, j = i+1, j-1 {
				perm[i], perm[j] = perm[j], perm[i]
			}
			flips++
			k = perm[0]
		}

		if flips > maxFlips {
			maxFlips = flips
		}
		if permCount%2 == 0 {
			checksum += flips
		} else {
			checksum -= flips
		}
		permCount++

		// Next permutation
		found := false
		for {
			if r == N {
				goto done
			}
			p0 := perm1[0]
			for i := 0; i < r; i++ {
				perm1[i] = perm1[i+1]
			}
			perm1[r] = p0
			count[r]--
			if count[r] > 0 {
				found = true
				break
			}
			r++
		}
		if !found {
			break
		}
	}

done:
	fmt.Printf("%d\nPfannkuchen(%d) = %d\n", checksum, N, maxFlips)
}
